spark/ingestion/src/main/scala/feast/ingestion/stores/redis/HashTypePersistence.scala

```diff
@@ -51,49 +57,104 @@ class HashTypePersistence(config: SparkRedisConfig) extends Persistence with Ser
       }
       .filter { case (k, _) =>
         // don't store entities & timestamp
-        !keyColumns.contains(k) && k != config.timestampColumn
+        !config.entityColumns.contains(k) && k != config.timestampColumn
       }
       .map { case (k, v) =>
         encodeKey(k) -> encodeValue(v, types(k))
       }
 
-    val timestamp = Seq(
+    val timestampHash = Seq(
       (
-        timestampField.getBytes,
+        timestampHashKey(config.namespace).getBytes,
         encodeValue(value.getAs[Timestamp](config.timestampColumn), TimestampType)
       )
     )
 
-    values ++ timestamp
+    val expiryTimestampHash = Seq(
+      (
+        expiryTimestampHashKey(config.namespace).getBytes,
+        encodeValue(expiryTimestamp, TimestampType)
+      )
+    )
+
+    values ++ timestampHash ++ expiryTimestampHash
   }
 
-  def encodeValue(value: Any, `type`: DataType): Array[Byte] = {
+  private def encodeValue(value: Any, `type`: DataType): Array[Byte] = {
     TypeConversion.sqlTypeToProtoValue(value, `type`).toByteArray
   }
 
-  def encodeKey(key: String): Array[Byte] = {
+  private def encodeKey(key: String): Array[Byte] = {
     val fullFeatureReference = s"${config.namespace}:$key"
     Hashing.murmur3_32.hashString(fullFeatureReference, StandardCharsets.UTF_8).asBytes()
   }
 
-  def save(
+  private def timestampHashKey(namespace: String): String = {
+    s"${config.timestampPrefix}:${namespace}"
+  }
+
+  private def expiryTimestampHashKey(namespace: String): String = {
+    s"${config.expiryPrefix}:${namespace}"
+  }
+
+  private def decodeTimestamp(encodedTimestamp: Array[Byte]): java.sql.Timestamp = {
+    new java.sql.Timestamp(Timestamp.parseFrom(encodedTimestamp).getSeconds * 1000)
+  }
+
+  override def save(
       pipeline: Pipeline,
       key: Array[Byte],
-      value: Map[Array[Byte], Array[Byte]],
-      ttl: Int
+      row: Row,
+      expiryTimestamp: java.sql.Timestamp
   ): Unit = {
-    pipeline.hset(key, value.asJava)
-    if (ttl > 0) {
-      pipeline.expire(key, ttl)
+    val value = encodeRow(row, expiryTimestamp).asJava
+    pipeline.hset(key, value)
+    if (!expiryTimestamp.equals(MAX_EXPIRED_TIMESTAMP)) {
+      pipeline.expireAt(key, expiryTimestamp.getTime / 1000)
     }
   }
 
-  def getTimestamp(
+  override def get(
       pipeline: Pipeline,
-      key: Array[Byte],
-      timestampField: String
-  ): Response[Array[Byte]] = {
-    pipeline.hget(key, timestampField.getBytes)
+      key: Array[Byte]
+  ): Response[util.Map[Array[Byte], Array[Byte]]] = {
+    pipeline.hgetAll(key)
+  }
+
+  override def storedTimestamp(
+      value: util.Map[Array[Byte], Array[Byte]]
+  ): Option[java.sql.Timestamp] = {
+    value.asScala.toMap
+      .map { case (key, value) =>
+        (key.map(_.toChar).mkString, value)
+      }
+      .get(timestampHashKey(config.namespace))
+      .map(value => decodeTimestamp(value))
   }
 
+  override def newExpiryTimestamp(
```

Comment:
This is not really persistence related method. Maybe move it to `RedisSinkRelation`?

Reply:
you can move `MAX_EXPIRED_TIMESTAMP` to package level